### PR TITLE
Fix context menu when long press on url and email link

### PR DESCRIPTION
**Pull request by b1tjoy**
### Details

When long press on URL/email link in comment, "Copy URL to clipboard" or "Copy e-mail to clipboard" does not show, this PR fix it by replace `LongPressGestureHandler` with `Pressable` in `PressableWithSecondaryInteraction`.

Because of the limitation of gesture responder system of react-native, only the inner `Pressable` in a nested `Pressable` tree can catch long press event, and because of the limitation of react-native-render-html, we can not pass necessary data to the inner `Pressable` to display context menu. So we use React Context API to let the inner pressable component display context menu in it's long press event handler.

But use `Pressable` cause some regressions, this PR fix the regressions too. Regressions need to notice:
1. Long press on image opens attachment modal on iOS and Android native, does not show context menu as we expected
2. Long press on URL link redirect to browser on iOS and Android native, does not show context menu as we expected
3. Long press on URL/email link opens 3D-Touch menu on iOS/Safari
4. Long press on code block shows nothing on iOS and Android native, does not show context menu as we expected
5. Long press on code block opens 3D-Touch menu on iOS/Safari
6. Long press on IOU card shows nothing on iOS and Android native, does not show context menu as we expected
7. Long press on IOU card opens 3D-Touch menu on iOS/Safari

Since we can pass props data to `IOUAction` in `ReportActionItem`, I use prop drilling to fix regressions with IOU card rather than Context API.

### Fixed Issues
$ https://github.com/Expensify/App/issues/8311   
PROPOSAL: https://github.com/Expensify/App/issues/8311#issuecomment-1312527754

### Tests
1. Send a mixed comment contains both regular text, URL and email
2. Long press on regular text in the message you just sent, verify it shows regular context menu
3. Long press on URL link in the message you just sent, verify it shows "Copy URL to clipboard" context menu
4. Press on URL link in the message you just sent, verify it opens URL with default browser
5. Long press on email link in the message you just sent, verify it shows "Copy email to clipboard" context menu
6. Send an image attachment
7. Long press on the image you just sent, verify it shows regular context menu
8. Press on the image you just sent, verify it shows attachment preview modal
9. Send a comment contains code block, line need to be long enough to make code block horizontally scrollable
10. Long press on the code block you just sent, verify it shows regular context menu
11. Use trackpad or finger to scroll code block and verify it can be scrolled horizontally
12. Use another account to request money from your current account, when you receive IOU action do the next step
13. Long press on Pay button or other area in IOU card, verify it shows regular context menu
14. Press on Pay button or other area in IOU card, verify it shows IOU details modal

- [x] Verify that no errors appear in the JS console

### QA Steps
1. Send a mixed comment contains both regular text, URL and email
2. Long press on regular text in the message you just sent, verify it shows regular context menu
3. Long press on URL link in the message you just sent, verify it shows "Copy URL to clipboard" context menu
4. Press on URL link in the message you just sent, verify it opens URL with default browser
5. Long press on email link in the message you just sent, verify it shows "Copy email to clipboard" context menu
6. Send an image attachment
7. Long press on the image you just sent, verify it shows regular context menu
8. Press on the image you just sent, verify it shows attachment preview modal
9. Send a comment contains code block, line need to be long enough to make code block horizontally scrollable
10. Long press on the code block you just sent, verify it shows regular context menu
11. Use trackpad or finger to scroll code block and verify it can be scrolled horizontally
12. Use another account to request money from your current account, when you receive IOU action do the next step
13. Long press on Pay button or other area in IOU card, verify it shows regular context menu
14. Press on Pay button or other area in IOU card, verify it shows IOU details modal

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The reviewer will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for expected offline behavior are in the `Offline steps` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [ ] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] I verified that if a function's arguments changed that all usages have also been updated correctly
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/103875612/203761672-c2c91994-e1a4-47b6-9975-e40326fc47ed.mp4

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/103875612/203761753-b3330e94-d15c-4ebb-8393-9f1cb9be62f6.mp4

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/103875612/203761826-91ed41a8-6f46-4c4d-adb4-d4d5117a4a69.mp4

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/103875612/203761899-71d4249f-205f-4e33-818e-9de0e2045c3f.mp4

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/103875612/203761970-7b7287c2-6fbb-49a4-bdd6-731b902fbf2d.mp4

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/103875612/203762008-f89dda7f-fe09-4eb7-aa66-a304d7beeced.mp4

</details>

### CODE REVIEW

**Comment by parasharrajat** [comment]
I will start the review today

File: src/components/AnchorForCommentsOnly/index.js [hunk 1]
```diff
@@ -1,11 +1,27 @@
...
+    ...anchorForCommentsOnlyPropTypes,
+    ...windowDimensionsPropTypes,
+};
+
+const AnchorForCommentsOnly = props => (
+    <BaseAnchorForCommentsOnly
+        onPressIn={() => props.isSmallScreenWidth && canUseTouchScreen() && ControlSelection.block()}
+        onPressOut={() => ControlSelection.unblock()}
+        // eslint-disable-next-line react/jsx-props-no-spreading
+        {...(_.omit(props, ['onPressIn', 'onPressOut']))}
```
Comment (parasharrajat) [inline src/components/AnchorForCommentsOnly/index.js:19]:
```suggestion
	// eslint-disable-next-line react/jsx-props-no-spreading
        {...props}
        onPressIn={() => props.isSmallScreenWidth && canUseTouchScreen() && ControlSelection.block()}
        onPressOut={() => ControlSelection.unblock()}
        
```

**Comment by parasharrajat** [comment]
It is much more involved than I thought.

File: src/components/ReportActionItem/IOUPreview.js [hunk 2]
```diff
@@ -133,8 +151,30 @@ const IOUPreview = (props) => {
...
+    };
+
     return (
-        <TouchableWithoutFeedback onPress={props.onPreviewPressed}>
+        <TouchableWithoutFeedback
+            onPress={props.onPreviewPressed}
+            onPressIn={() => props.isSmallScreenWidth && canUseTouchScreen() && ControlSelection.block()}
+            onPressOut={() => ControlSelection.unblock()}
+            onLongPress={showContextMenu}
+        >
```
Comment (parasharrajat) [inline src/components/ReportActionItem/IOUPreview.js:177]: I didn't expect these to be changed. But let me test these first.

**Review by parasharrajat — COMMENTED** [review]
Started review.

**Comment by parasharrajat** [comment]
Sorry for the delay on this. I am currently away. I will actively review it when I am back.

**Comment by marcochavezf** [comment]
Hi @b1tjoy, could you fix the conflicts?

**Review by marcochavezf — CHANGES_REQUESTED** [review]
Thanks for the changes @b1tjoy, the code LGTM so far! I'm leaving a few comments:

**Comment by marcochavezf** [comment]
Hi @parasharrajat, let me know if you're actively reviewing this PR, happy to help to review/test it in spare cycles if not.

**Comment by marcochavezf** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>



</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>



</details>

<details>
<summary>Desktop</summary>



</details>

<details>
<summary>iOS</summary>



</details>

<details>
<summary>Android</summary>



</details>

**Comment by marcochavezf** [comment]
Hi @b1tjoy, could you fix again the conflicts? I was OOO, apologies for the delay here. I'm going to continue reviewing it asap once the conflicts are resolved 🙇🏽

**Review by marcochavezf — APPROVED** [review]
LGTM, I tested all possible scenarios, and seems everything is working fine 👍🏽 Thanks @b1tjoy for all the effort here!

**Comment by mollfpr** [comment]
Heyy, coming from this https://github.com/Expensify/App/issues/14301 issue. Our contributors find this PR introduces the issue, specifically the usage of `Pressable` in the `BasePreRenderer` component when hovered, the cursor has the pointer style instead of text. To fix it, we replace the `Pressable` with `TouchableWithoutFeedback` and have discussed it in [Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1674844591879669).

Thank you!

**Comment by sobitneupane** [comment]
Going through the issue and PR, it looks like the changes made in this PR is intended to solve the issue in native and mWeb or particularly touch screen devices. [This issue](https://github.com/Expensify/App/issues/17487) reported suggest that the feature of long press should not be available for Web and Desktop. So, we prevented the Long Press Context Menu for Web and Desktop with [these changes](https://github.com/Expensify/App/pull/17782/files)

**Pull request merged**